### DISCUSSION TITLE
Make Taskbar's overlap resize compatible with the built-in macOS double click to maximize

### DISCUSSION THREAD

**Discussion by sevban** [lawand-dot-io/taskbar#94] [Feature requests]
**Describe the bug**
When Taskbar is running double clicking on a title bar doesn't restore previous window size.

**To Reproduce**
Open Taskbar advanced settings and make sure "Auto-resize windows when they overlap Taskbar" setting is active.
Open any app or for example Chrome.
Drag from the corner and change its window size.
Now double click the title bar and it will maximize the window with a zoom effect.
Double click again on the titlebar.
It should restore the previous window size but that setting prevents this.

**Expected behavior**
Window should restore to its previous size.

**Desktop (please complete the following information):**
 - Taskbar version 1.2.5
 - OS: Sequoia 15.0
 - M3 Pro
 
I know this happens because Taskbar resizes the window to prevent overlap but I think it there should be a more elegant solution to overlapping to prevent this "bug".

**Comment by lawand-dot-io** [comment]
Hello. Thanks for mentioning this. I don't see a way of how this could be technically possible since this is an unavoidable interaction between Taskbar resizing overlapping windows and macOS's title bar double click. I will convert this into a feature request and keep it open for now because maybe I would learn something new in the future to know how to stop this behavior

**Comment by marcellus-trixus** [comment]
+1 as this problem gives me a permanent headache.

I am not sure how the overlapping size correction works as this repo does not contain code. But I assume that there must be some kind of a callback from the OS that a window was resized and OS calls to get and set the size of a window. 

I am no expert in the field of macos internals, but If both assumptions are true, it **might** be fixed by doing the resize twice.

1. Get a resize callback (`[Size1] -> [Size2]`)
2. Calculate correct size `[Size3]` but do not apply it yet
3. Set the window's size back (again) to the original size before maximize action (`Size1`)
4. Set the window's size to the correct size `[Size3]`

This could theoretically make the MacOS think that the previous "before maximizing" size was the original `[Size1]` instead of `[Size2]`. Maybe I am totally wrong, but it might be worth the shot :-)

**Reply by lawand-dot-io** [reply]
Thanks for this analysis. I think it doesn't work like this since, if I understand correctly, when you double click a window's title macOS checks the current window size. If the window doesn't take up the whole screen then make it take up the whole screen. If it already does then return to the previous (last) size. So in my case, double clicking when the window takes up the whole screen minus taskbar will always make it take up the whole screen then taskbar notices that and resize it so it it takes up the whole screen minus taskbar.

I think the solution to this problem is that when taskbar gets a resize event, if the window was resized from "whole screen minus taskbar" to "whole screen" then in this case instead of setting the size to "whole screen minus taskbar" it would set it to the original size. This works in case I can always keep track of window sizes, so I will have to look into this.

Thanks again for the comment and I will keep this in mind. Cheers
